### DISCUSSION TITLE
PrismaClientValidationError: Argument is missing - How to resolve?

### DISCUSSION THREAD

**Discussion by delfincortesb** [prisma/prisma#20216] [Q&A]
I'm encountering a PrismaClientValidationError when trying to update a record using the Prisma ORM. The error specifically states that the argument 'companies' is missing in the 'prisma.documents_counter.update()' invocation. Here's the error message I receive:

```js
PrismaClientValidationError: Invalid `prisma.documents_counter.update()` invocation:

{
  where: {
    company_id: 1
  },
  data: {
    estimate_count: {
      increment: 1
    },
+   companies: {
+     create: companiesCreateWithoutEstimatesInput | companiesUncheckedCreateWithoutEstimatesInput,   
+     connectOrCreate: companiesCreateOrConnectWithoutEstimatesInput,
+     connect: companiesWhereUniqueInput
+   }
  }
}

Argument 'companies' is missing.
```

Here are some details and code snippets related to my setup:

1. I'm using Prisma version 5.0.0
2. The error occurs when I try to update a record in the 'documents_counter' table.
3. The 'documents_counter' table has a relation to the 'companies' table, defined in the Prisma schema.
4. I'm using the 'update' method provided by Prisma to modify the 'documents_counter' record.
5. The error points to the missing 'companies' argument in the update invocation.

Here's an example code snippet that illustrates the structure of my update operation:

```js
const estimate = await prisma.documents_counter.update({
  where: {
    company_id: 1, // Replace with the actual company_id
  },
  data: {
    estimate_count: {
      increment: 1,
    },
    // Missing 'companies' argument here
  },
});
```

I have reviewed the Prisma schema, and the 'companies' field is correctly defined in the 'documents_counter' model. However, when I try to include the 'companies' field in the update operation, I receive the 'Argument 'companies' is missing' error.

Here's some part of the prisma schema:
```prisma
model estimates {
  id                     Int               @id(map: "invoices_duplicate_pkey") @default(autoincrement())
  number                 String?
  customer_id            Int
  created_at             DateTime          @db.Timestamp(6)
  last_modified_date     DateTime?         @db.Timestamp(6)
  exp_at                 DateTime          @default(now()) @db.Timestamp(6)
  adjustment_total       Float?            @db.Real
  adjustment_description String?
  sub_total              Float             @db.Real
  discount_total         Float?            @db.Real
  tax_total              Float             @db.Real
  total                  Float             @db.Real
  tax_inclusive          Boolean           @default(true)
  warehouse_id           Int
  company_id             Int
  companies              companies         @relation(fields: [company_id], references: [id], onDelete: NoAction, onUpdate: NoAction)
  customers              customers         @relation(fields: [customer_id], references: [id], onDelete: NoAction, onUpdate: NoAction)
  warehouses             warehouses        @relation(fields: [warehouse_id], references: [id], onDelete: NoAction, onUpdate: NoAction)
  estimates_items        estimates_items[]

  @@schema("public")
}
model companies {
  id                Int                 @id @default(autoincrement())
  name              String
  rnc               String?
  address           String?
  city              String?
  province          String?
  cedula            String?
  username          String              @unique @default("")
  chart_of_accounts chart_of_accounts[]
  companies_users   companies_users[]
  documents_counter documents_counter?
  documents_ncfs    documents_ncfs[]
  estimates         estimates[]
  invoices          invoices[]
  payment_methods   payment_methods[]
  taxes             taxes[]
  warehouses        warehouses[]

  @@schema("public")
}
model documents_counter {
  id              Int       @id(map: "transaction_counters_pkey") @default(autoincrement())
  company_id      Int       @unique
  invoice_prefix  String?   @default("FAC")
  invoice_count   Int?      @default(0)
  estimate_prefix String?   @default("EST")
  estimate_count  Int?      @default(0)
  companies       companies @relation(fields: [company_id], references: [id], onUpdate: NoAction)

  @@schema("public")
}

```

**Comment by ludralph** [comment]
Hi @delfincortesb 👋 

Thank you for raising this question.

> when I try to include the 'companies' field in the update operation, I receive the 'Argument 'companies' is missing' error.

Can you please clarify if you are trying to include the `companies` relation in the update query as shown in the code below?

```js
const res = await prisma.documents_counter.update({
  where:{
    company_id: 2,
  },
  data:{
    estimate_count: {
      increment: 1,
    }, 
  },
  include: {
    companies: true
  }
})
```
Or are you trying to perform a nested update on the companies relation as shown in the code snippet below:
```js
const res = await prisma.documents_counter.update({
  where:{
    company_id: 2,
  },
  data:{
    estimate_count: {
      increment: 1,
    }, 
    companies:{
        create:{
          name:"test",
          username:"gab",
        }
    }
  },
})

**Reply by nurul3101** [reply]
@delfincortesb Did you had a chance to look at Raphael's response? Please let us know in case you still need assistance.
If not, please consider marking the discussion as answered/closed.

**Reply by dengel29** [reply]
It's unclear who Raphael is, please refer to their username in the future. Thanks

**Comment by LeandroVCastro** [comment]
This bug still happening to me.
I'm using client v5.2

**Reply by hyperVineet** [reply]
Something similar happened with me... i was trying to store a null value in a column of type String[]... Turns out that doesn't work well in prisma... But prisma was giving incorrect logs saying `Argument is missing` like this pr... Storing [] instead of null solved the issue...

**Reply by LeandroVCastro** [reply]
The same happened to me. I realized this is not related to any argument missing.
So maybe to fix this issue is just improve this error message.

On my case I can't remember exactly what was going, but I'm sure wasn't related to argument missing.

**Reply by Merkie** [reply]
I just found this thread from the error: Argument `set` is missing.

Turns out one of the objects I was using in my .createMany statement had an object instead of an array as one of its properties when it was only supposed to be an array. I wish the error handling on the client could indicate this; it took me parsing through the entire stack to figure out what was wrong, moral of the story is validate all data before sending it along to prisma because at the end of the day prisma is not a data validator 🤷‍♂️ 

funny seeing this thread so active though lol

edit: *and just to be clear, I had no properties named set in my schema, this seems to be an internal error making its way to the surface*

**Reply by vjpr** [reply]
> Turns out one of the objects I was using in my .createMany statement had an object instead of an array as one of its properties when it was only supposed to be an array.

Same with me. Was passing `{}` to `Json[]`. Changed to `[{}]`

**Comment by bistacos** [comment]
+1, When I accidentally try to set a non-null field to `null`, I get an unrelated "Argument `X` is missing" rather than an error telling me that I've got an invalid value in my insert statement. This sent me barking up the wrong tree for a while; good opportunity for better error handling here.

**Comment by CodrinSocol** [comment]
+1. I have encountered the same issue. It seems that I had the same issue as @Merkie described above. I was passing a JSON Object instead of a JSON Array and that's why Prisma complained. It would have been nice to have the error handling indicate this instead of `Argument set is missing`.

**Comment by strongui** [comment]
Anyone else stumbling onto this. You will also get this error if you try to search for `null` on a **non-nullable field**. Took me a second to figure out, because the message doesn't point you in the right direction.

**Comment by LeandroVCastro** [comment]
I was looking for the file that throws this error message, and seems that is it: https://github.com/prisma/prisma/blob/main/packages/client/src/runtime/core/errorRendering/applyValidationError.ts

This error apparently is throwing on line 306: 
> return `Argument \`${colors.green(argumentName)}\` is missing.`

I think maybe this message could be changed to something like 
> "Argument 'anything' is not nullable neither has a default value, but it is not setted"

I could do it, but I would like to discuss about this before. 
What about this, guys?

**Comment by dengel29** [comment]
Yeah I'm facing this currently and the error message is leading me nowhere. Running my upsert works fine in Apollo Studio (input and object types + relevant files + functions below) but are throwing this error when running e2e tests with jest. For what it's worth I'm using GraphQL and NestJS

The error message is:

``` 
Invalid `this.prismaService.foodJournal.create()` invocation in
    /Users/dan/code/api.nuli.app/src/domains/foodJournal/service.ts:26:56

  25 async create(data: UpsertFoodJournalInput): Promise<FoodJournal> {
    → 26     const f = await this.prismaService.foodJournal.create({
               data: {
                 userId: "bb0a4539-edd6-465b-acfb-6e9b095cdfe0",
                 id: null,
                 note: "Hellllllo",
                 date: new Date("2010-02-09T16:00:00.000Z"),
             +   user: {
             +     create: UserCreateWithoutFoodJournalsInput | UserUncheckedCreateWithoutFoodJournalsInput,
             +     connectOrCreate: UserCreateOrConnectWithoutFoodJournalsInput,
             +     connect: UserWhereUniqueInput
             +   }
               }
             })
```

But if we look at the referenced line, here's what I have in `service.ts`:

```ts
 async create(data: UpsertFoodJournalInput): Promise<FoodJournal> {
        return await this.prismaService.foodJournal.create({ data });
    }
```



Here are the relevant files:

in prisma.schema
```ts
model FoodJournal {
  id        String   @id @default(dbgenerated("(uuid())")) @db.Char(36)
  note      String?  @db.Text
  userId    String   @db.Char(36)
  date      DateTime @db.DateTime(0)
  createdAt DateTime @default(now()) @db.DateTime(0)
  updatedAt DateTime @default(now()) @updatedAt @db.DateTime(0)
  user      User     @relation(name: "userFoodJournal", fields: [userId], references: [id], onDelete: Cascade)

  @@index([userId], map: "userId")
  @@map("food_journal")
}
```

in types.ts
```ts
@InputType()
export class UpsertFoodJournalInput {
    @Field(() => ID)
    @IsUUID()
    userId: FoodJournal['userId'];

    @Field(() => ID, { nullable: true })
    @IsOptional()
    @IsUUID()
    id?: string;

    @Field(() => String, { nullable: true })
    @IsOptional()
    note?: FoodJournal['note'];

    @Field(() => Date, { middleware: [dateMiddleware] })
    date: FoodJournal['date'];
}

@ObjectType('FoodJournal', {
    description: 'FoodJournal',
})
export class FoodJournalDTO implements FoodJournal {
    @Field(() => ID)
    @IsUUID()
    id: FoodJournal['id'];

    @Field(() => ID)
    @IsUUID()
    userId: FoodJournal['userId'];

    @Field(() => String, { nullable: true })
    note: FoodJournal['note'];

    @Field(() => Date, { middleware: [dateMiddleware] })
    date: FoodJournal['date'];

    @Field(() => Date, { middleware: [dateMiddleware] })
    createdAt: Date;

    @Field(() => Date, { middleware: [dateMiddleware] })
    updatedAt: Date;
}
```

in test.spec.ts
```ts
describe('Domains > Food Journal', () => {
    let request: E2ERequest;
    let prismaService: PrismaService;
    let app: INestApplication;
    let userId = '';
  const defaultData = {
      date: new Date(),
      note: 'Hellllllo',
      id: null,
  };

    const { beforeAllHook, afterAllHook } = getE2EHooks('e2e-foodJournals');
    beforeAll(async () => {
        ({ request, prismaService, app } = await beforeAllHook());

        const user = await prismaService.user.findUniqueOrThrow({
            where: { email: 'user1@user.io' },
        });
        userId = user.id;
    });

    afterAll(async () => await afterAllHook());

    beforeEach(async () => {
        await app.get(RedisProvider);
    });

    describe('Food Journal Resolver', () => {
        it('should create a food journal for a given date', async () => {
            const variables = {
                foodJournal: {
                    ...defaultData,
                    userId,
                },
            };

            const { createdFoodJournal } = await request<
                { createdFoodJournal: FoodJournalDTO },
                { foodJournal: UpsertFoodJournalInput }
            >({
                document:  gql`
                  mutation UpsertFoodJournal($foodJournal: UpsertFoodJournalInput!) {
                      upsertFoodJournal(foodJournal: $foodJournal) {
                          id
                          userId
                          note
                          date
                          createdAt
                          updatedAt
                      }
                  }
              `;`,
                variables,
            });

            expect(createdFoodJournal).not.toBeUndefined();
        });
    });
});
```

**Comment by jannikhst** [comment]
I encountered the problem when I performed a batch update. The validation error occurred in the ~100 command within the batch, but the first command was displayed to cause the error. Took me a bit to find the issue.

**Comment by kassiogluten** [comment]
i had a similar problem, and solved by updating schema to JSON[]

**Comment by hamzathul** [comment]
I had the same issue, and I went through the data given to Prisma carefully, and I found that a field was undefined and fixed it. Make sure the input is validated properly.
